File: src/main/java/net/gegy1000/terrarium/server/map/cover/generator/layer/SelectWeightedLayer.java
package net.gegy1000.terrarium.server.map.cover.generator.layer;

import net.minecraft.world.gen.layer.GenLayer;
import net.minecraft.world.gen.layer.IntCache;

public class SelectWeightedLayer extends GenLayer {
    private final Entry[] entries;
    private final int totalWeight;

    public SelectWeightedLayer(long seed, Entry... entries) {
        super(seed);
        this.entries = entries;

        int totalWeight = 0;
        for (Entry entry : entries) {
            totalWeight += entry.weight;
        }
        this.totalWeight = totalWeight;
    }

    @Override
    public int[] getInts(int areaX, int areaY, int areaWidth, int areaHeight) {
        int[] result = IntCache.getIntCache(areaWidth * areaHeight);
        for (int z = 0; z < areaHeight; z++) {
            for (int x = 0; x < areaWidth; x++) {
                this.initChunkSeed(areaX + x, areaY + z);
                result[x + z * areaWidth] = this.getSelection(this.nextInt(this.totalWeight));
            }
        }
        return result;
    }

    private int getSelection(int weight) {
        int currentWeight = 0;
        for (Entry entry : this.entries) {
            currentWeight += entry.weight;
            if (currentWeight >= weight) {
                return entry.value;
            }
        }
        return 0;
    }

    public static class Entry {
        private final int value;
        private final int weight;

        public Entry(int value, int weight) {
            this.value = value;
            this.weight = weight;
        }
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/map/glob/generator/layer/SelectWeightedLayer.kt
package net.gegy1000.terrarium.server.map.glob.generator.layer

import net.minecraft.world.gen.layer.GenLayer
import net.minecraft.world.gen.layer.IntCache

class SelectWeightedLayer(seed: Long, vararg val entries: Entry) : GenLayer(seed) {
    val totalWeight: Int

    init {
        this.totalWeight = this.entries.sumBy { it.weight }
    }

    override fun getInts(areaX: Int, areaY: Int, areaWidth: Int, areaHeight: Int): IntArray {
        val result = IntCache.getIntCache(areaWidth * areaHeight)
        for (z in 0..areaHeight - 1) {
            for (x in 0..areaWidth - 1) {
                this.initChunkSeed((areaX + x).toLong(), (areaY + z).toLong())
                val selectedWeight = this.nextInt(this.totalWeight)
                var currentWeight = 0
                for ((id, weight) in this.entries) {
                    currentWeight += weight
                    if (currentWeight >= selectedWeight) {
                        result[x + z * areaWidth] = id
                        break
                    }
                }
            }
        }
        return result
    }

    data class Entry(val id: Int, val weight: Int)
}
